### PR TITLE
Residual object

### LINKED ISSUE #13609: ResidualObject?
## Reason
There are some significant code duplications in `Kernel`, `BoundaryCondition`, `InterfaceKernel`, `DGKernel`, `Constraint`, `NodalKernel`, for example, the code for save-in and tagging capabilities.

## Design
A suggestion is to create a common base class like `ResidualObject` (possibly derived from `MooseObject` and several interface classes), where we can put the common codes in. This also provides us a capability to check if an object is changing residual or Jacobian during the run-time.

## Impact
A suggestion for better code.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #13609.

This will be useful for the eigen system cleanups.

Note that the second comment went a little further by replacing `computeOffDiagJacobian(unsigned jvar)` with `computeOffDiagJacobian(MooseVariableFEBase & jvar)`. It does not change the interface of `computeQpOffDiagJacobian`. Thus this PR may break applications if applications override `computeOffDiagJacobian`. It should be pretty rare though.

### CODE REVIEW

**Comment by YaqiWang** [comment]
I have to give up the further cleanup, simple because too much objects affected... We can create a script to facilitate that in the future.

**Comment by YaqiWang** [comment]
This showed good on my mac and on sawtooth. Should be ready for review. @fdkong or @lindsayad.

**Comment by YaqiWang** [comment]
One test is failing in debug mode due to an assertion. The two variables are defined on the original mesh and displaced mesh. Somehow the size of the local Jacobian is zero. What puzzles me is that this does not fail before. Any idea?

**Comment by fdkong** [comment]
Looks like you can not totally get of that API because some animals are still using it. It is better to just deprecate it.

**Comment by loganharbour** [comment]
I'm a fan of this. I will look into it once tests are passing.

File: framework/include/base/ResidualObject.h
Comment (fdkong) [inline framework/include/base/ResidualObject.h:65]: It might be even better to use `MooseVariableBase` to cover all cases

**Review by fdkong — CHANGES_REQUESTED** [review]
It looks really good for me. If it passed all tests, I would just get it in

**Comment by fdkong** [comment]
> One test is failing in debug mode due to an assertion. The two variables are defined on the original mesh and displaced mesh. Somehow the size of the local Jacobian is zero. What puzzles me is that this does not fail before. Any idea?

It look likely a variable is not defined on that interface or element but `computeJacobian` is still involved. We could check if the number of DoFs is nonzero before we call `computeJacobian`

File: framework/include/base/ResidualObject.h
Comment (YaqiWang) [reply framework/include/base/ResidualObject.h:65]: That will be too aggressive.

File: framework/src/bcs/IntegratedBC.C
Comment (lindsayad) [inline framework/src/bcs/IntegratedBC.C:142]: Did you read this comment before you made this change?

Comment (YaqiWang) [reply framework/src/bcs/IntegratedBC.C:142]: ah, no. lol.

Comment (lindsayad) [reply framework/src/bcs/IntegratedBC.C:142]: It is telling you the exact problem that happened

Comment (lindsayad) [reply framework/src/bcs/IntegratedBC.C:142]: `_sys` is the displaced nonlinear system if `use_displaced_mesh` for that object is `true`, else it is the undisplaced nonlinear system

Comment (YaqiWang) [reply framework/src/bcs/IntegratedBC.C:142]: There two other places to fix: `NodeFaceConstraint.C ` and `GapHeatTransfer.C`.

File: framework/include/base/ResidualObject.h
Comment (lindsayad) [inline framework/include/base/ResidualObject.h:36]: Don't need this here

File: framework/include/base/ResidualObject.h
Comment (lindsayad) [reply framework/include/base/ResidualObject.h:73]: We want the parameter here to be `unsigned int jvar` instead of `MooseVariable.*`. Having this parameter as `MooseVariable.*` can more easily lead to errors like the one you encountered in the displaced integrated boundary condition test. This will require modifying `ComputeFullJacobianThread::computeFaceJacobian` but that's ok

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by loganharbour** [comment]
While there is general agreement on using `ResidualObject`, there is a desired way of satisfying the other issue of accessing tagged objects in `NonlinearEigenSystem`. See https://github.com/idaholab/moose/issues/16125. I will work on this today.

**Comment by YaqiWang** [comment]
@lindsayad I think we should keep `void computeOffDiagJacobian(MooseVariableFEBase & /*jvar*/) ` instead of `void computeOffDiagJacobian(unsigned int /*jvar*/)`. The reason is made clearer in `Kernel::computeOffDiagJacobian`, where we need to check the variable type, normal variable or array variable. Basically a full variable reference provides more information like number of components than only the variable number. I do not worry about using the variables in a wrong system because it is handled in the framework. Potentially we can move some common codes in `Kernel::computeOffDiagJacobian` into `ResidualObject::computeOffDiagJacobian`. But I am not going to pursue that in this PR. I did this PR for the eigenvalue executioner cleanup.

**Comment by lindsayad** [comment]
>  I do not worry about using the variables in a wrong system because it is handled in the framework.

This is the framework. And clearly it's not handled as evinced by the earlier test error

**Comment by lindsayad** [comment]
> The reason is made clearer in `Kernel::computeOffDiagJacobian`, where we need to check the variable type, normal variable or array variable.

You already have this call in the code you referenced:

```c++
    // This (undisplaced) jvar could potentially yield the wrong phi size if this object is acting
    // on the displaced mesh
    auto phi_size = _sys.getVariable(_tid, jvar.number()).dofIndices().size();
```

which we know we need to avoid the error seen with the integrated bcs. So save-off a reference to the variable here in the `Kernel.C` code and then use it within the method instead of using the `jvar` argument

**Comment by YaqiWang** [comment]
You mean save off the pointers of all primal variables and use `unsigned int jvar` to retrieve one? It is then possibly better to move the `use_displaced_mesh` parameter down into residual object. I want this PR a minimum change for introducing `ResidualObject`. We can do that later, can we?

**Comment by YaqiWang** [comment]
I noticed a few other things in this PR: `IntegratedBC` is using a name `computeJacobianBlock` instead of `computeOffDiagJacobian`; IMO `ODEKernel::computeJacobian()` should be implemented in `ScalarKernel::computeJacobian()`; `NodalConstraint` has different interface for `computeResidual` and `computeJacobian`, etc. Not sure if I should submit a separate issue for these. Just to let you know.

**Comment by lindsayad** [comment]
> You mean save off the pointers of all primal variables and use `unsigned int jvar` to retrieve one? It is then possibly better to move the `use_displaced_mesh` parameter down into residual object. I want this PR a minimum change for introducing `ResidualObject`. We can do that later, can we?

No. This is the code right now:
```c++
void
Kernel::computeOffDiagJacobian(MooseVariableFEBase & jvar)
{
  auto jvar_num = jvar.number();
  if (jvar_num == _var.number())
    computeJacobian();
  else
  {
    prepareMatrixTag(_assembly, _var.number(), jvar_num);

    // This (undisplaced) jvar could potentially yield the wrong phi size if this object is acting
    // on the displaced mesh
    auto phi_size = _sys.getVariable(_tid, jvar.number()).dofIndices().size();
```
This is the proposed code:
```
void
Kernel::computeOffDiagJacobian(unsigned int jvar_num)
{
  const auto & jvar = _sys.getVariable(_tid, jvar_num);

  if (jvar_num == _var.number())
    computeJacobian();
  else
  {
    prepareMatrixTag(_assembly, _var.number(), jvar_num);

    // This (undisplaced) jvar could potentially yield the wrong phi size if this object is acting
    // on the displaced mesh
    auto phi_size = jvar.dofIndices().size();
```
The rest of the method would be totally unchanged. You were already talking about deleting an `offDiagJacobian` API. We're just trying to make sure we delete the right one.

**Comment by YaqiWang** [comment]
I can argue we pass in the right variable within `ComputeFullJacobianThread`. `SystemBase::getVariable` call is not so light. I still feel the interface using the variable is the right one.

**Comment by YaqiWang** [comment]
Note this:
```
Kernel::computeOffDiagJacobian(unsigned int jvar)
{
  mooseDeprecated("The computeOffDiagJacobian method signature has changed. Developers, please "
                  "pass in a MooseVariableFEBase reference instead of the variable number");
```
Someone already thought of this before.

**Comment by lindsayad** [comment]
I'm aware so who more qualified to give an opinion than someone who has done it both ways?

**Comment by lindsayad** [comment]
> `SystemBase::getVariable` call is not so light. I still feel the interface using the variable is the right one.

Have you seen it show up in profiling?

**Comment by YaqiWang** [comment]
Nope and it is probably not important. But that is an indicator how the code should be designed. The loops can have a hold of two sets of variables and pass in the correct one to the residual objects.

**Comment by lindsayad** [comment]
So you still want to do a container query

**Comment by lindsayad** [comment]
If you provide a clean implementation with `computeOffDiagJacobian(MooseVariable.*)` which does not require any `_sys.getVariable` calls in the method in order to get the correct variable, then I won't shut the door on that API. But it has to be in this PR. I won't accept this as-is with `_sys.getVariable` calls in `computeOffDiagJacobian(MooseVariable.*)` because that simply exposes that we've set up a dangerous way of communicating coupling information which is liable to break when doing coupled undisplaced/displaced simulations.

You've put up a good PR @YaqiWang that I've gone to bat for you for. You've even exposed flaws in a design that I myself created, as you pointed out. So let's use this excellent opportunity that you've created to make things as beautiful as possible.

**Comment by YaqiWang** [comment]
This is a good argument. @lindsayad Not a problem to me. I think your argument is good.

**Comment by YaqiWang** [comment]
Noticed that: `defineLegacyParams(ResidualObject);` in ResidualObject.C was not deleted.

**Comment by lindsayad** [comment]
@YaqiWang can you explain why you removed `computeQpOffDiagJacobian(unsiged int)` from `ScalarKernel`?

**Comment by lindsayad** [comment]
> @YaqiWang can you explain why you removed `computeQpOffDiagJacobian(unsiged int)` from `ScalarKernel`?

I guess this was actually removed from `ODEKernel`. I can see that we only actually do off-diagonal Jacobian computation (looking at `ODEKernel::computeOffDiagJacobian`) for scalar variables, so maybe that was your reasoning...which I think is good reasoning.

**Comment by lindsayad** [comment]
Yea to your earlier point, I don't really understand why we have `ScalarKernel` and `ODEKernel` separated, but I guess `GlobalStrain` does inherit from `ScalarKernel` as opposed to `ODEKernel` and it implements much different `computeResidual` and `computeJacobian`

**Comment by YaqiWang** [comment]
I think we should clean up that scalar kernel. But we can do it in a separate PR.

File: framework/src/base/ResidualObject.C
Comment (YaqiWang) [inline framework/src/base/ResidualObject.C:61]: What the prepare shapes is doing? Are the two kernels having the same thing thus we could be duplicating computation here?

File: framework/src/kernels/VectorKernel.C
Comment (YaqiWang) [inline framework/src/kernels/VectorKernel.C:102]: const unsigned int?

File: framework/src/loops/ComputeFullJacobianThread.C
Comment (YaqiWang) [inline framework/src/loops/ComputeFullJacobianThread.C:63]: Oh, I see here. We are not introducing anything new then.

File: framework/include/constraints/Constraint.h
Comment (YaqiWang) [inline framework/include/constraints/Constraint.h:25]: missing NeighborResidualObject source in this commit?

Comment (lindsayad) [reply framework/include/constraints/Constraint.h:25]: Yea it's in a fixup commit (trying to avoid force pushing to your repo)

Comment (lindsayad) [reply framework/include/constraints/Constraint.h:25]: https://github.com/idaholab/moose/pull/16117/commits/85350e9d2c7f3d54b5d03424d280968f0648ca90

File: framework/include/base/NeighborResidualObject.h
Comment (YaqiWang) [inline framework/include/base/NeighborResidualObject.h:3]: This should be merged into previous commit. The commit message is not right.

File: framework/src/loops/ComputeFullJacobianThread.C
Comment (lindsayad) [reply framework/src/loops/ComputeFullJacobianThread.C:63]: Right, this is just introducing more encapsulation. I feel like if we have a non-const `subProblem()` getter in our residual objects, we are doing something wrong

File: framework/src/kernels/VectorKernel.C
Comment (lindsayad) [reply framework/src/kernels/VectorKernel.C:102]: This is to prevent some deviant from coming into the method body and reassigning `jvar_num`.

File: framework/include/constraints/Constraint.h
Comment (YaqiWang) [reply framework/include/constraints/Constraint.h:25]: Force push should be fine. I stopped working on this branch. It is yours.

File: framework/include/base/NeighborResidualObject.h
Comment (lindsayad) [reply framework/include/base/NeighborResidualObject.h:3]: You've never used `git commit --fixup` and `git rebase --autosquash` ? It is easy to autosquash all these commits later. If you give me permission to force push willy nilly to this branch, then I won't push up `fixup` commits anymore

**Comment by YaqiWang** [comment]
Really nice. I think we should add some inline comment before `ResidualObject::computeOffdiagJacobian` to give reasons why we design the interface in this way.

File: framework/include/base/ResidualObject.h
Comment (YaqiWang) [inline framework/include/base/ResidualObject.h:67]: It will be nice to add some docs here to explain why we should to pass variable number here. Note this line of doxygen comment is not accurate anymore.

File: framework/include/base/NeighborResidualObject.h
Comment (YaqiWang) [reply framework/include/base/NeighborResidualObject.h:3]: Oh, you cannot do force push?

Comment (YaqiWang) [reply framework/include/base/NeighborResidualObject.h:3]: My repository is public. I thought anyone can force push to my branches. If not, where should set this up?

Comment (YaqiWang) [reply framework/include/base/NeighborResidualObject.h:3]: Oh maybe feel free to create a new PR with your own branch.

Comment (lindsayad) [reply framework/include/base/NeighborResidualObject.h:3]: I can. Was just not doing it at first out of courtesy to you. But now that you've said this branch is all mine, I'll force push :+1:

**Comment by lindsayad** [comment]
Patches:

- [x] Griffin
- [ ] THM
- [ ] Mastodon

**Comment by YaqiWang** [comment]
@fdkong I guess this can be merged.

**Comment by fdkong** [comment]
> @fdkong I guess this can be merged.

It is still failing. We need to force merge failing things for integration?

**Comment by loganharbour** [comment]
Please don't merge until we at least have fixes ready to be merged up for the affected apps.

**Comment by lindsayad** [comment]
See https://github.com/idaholab/moose/pull/16117#issuecomment-725125095. This will probably be one to two days work to get this all integrated, but I think it's well worth it. It's impossible to be backwards compatible when changing APIs that currently take a non-const reference to taking a const reference (at least I haven't thought of a way to do it)

**Comment by loganharbour** [comment]
> See [#16117 (comment)](https://github.com/idaholab/moose/pull/16117#issuecomment-725125095). 

I saw - I made the comment because I believe you're missing one or two

**Comment by lindsayad** [comment]
> I saw - I made the comment because I believe you're missing one or two

Which ones? They're failures independent of THM/Griffin?

**Comment by loganharbour** [comment]
> They're failures independent of THM/Griffin?

Disregard, I failed to note the dependency

**Comment by YaqiWang** [comment]
@lindsayad Looks like that we still need to check the following three:
- [ ] Pronghorn
- [ ] Relap-7
- [ ] Sockeye

I can help Pronghorn and Relap-7.

**Comment by loganharbour** [comment]
> @lindsayad Looks like that we still need to check the following three:

Those are good. Look at the conversation above about THM/Griffin dependency.

**Comment by YaqiWang** [comment]
Where is the conversation?

**Review by loganharbour — APPROVED** [review]
For @lindsayad's commits

**Review by lindsayad — APPROVED** [review]
Removing my un-approval. @fdkong you want to look once more? You are the only one with outstanding requested changes

**Review by lindsayad — APPROVED** [review]
Removing my un-approval. @fdkong you want to look once more? You are the only one with outstanding requested changes

**Comment by fdkong** [comment]
> Removing my un-approval. @fdkong you want to look once more? You are the only one with outstanding requested changes

I am looking at it now

**Review by fdkong — APPROVED** [review]

**Comment by fdkong** [comment]
OK, merged. @lindsayad Could you follow up the integration process?

**Comment by YaqiWang** [comment]
Congratulation! ;-)

**Comment by lindsayad** [comment]
> OK, merged. @lindsayad Could you follow up the integration process?

Yes

**Pull request merged**
